src/data/sonata_data.cpp

```diff
@@ -160,11 +161,11 @@ void SonataData::record_data(double step) {
         local_position += kv.second->get_num_elements();
     }
     current_step_++;
-    last_position_ += total_elements_;
-    last_step_recorded_ += reporting_period_;
+    previous_position_ += total_elements_;
+    previous_step_recorded_ += reporting_period_;
 
     if (current_step_ == steps_to_write_) {
-        write_data();
+        write_data(report_buffer_, current_step_);
```

Comment:
Why not directly `flush()` here and in line 198?

Reply:
Done!